### ISSUE TITLE
use python 2.6's streaming zipfile support instead of our own zipstream module

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4035]
|<img alt="kmike's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| kmike reported|
|-|-|
|Trac ID|trac#4035|
|Type|enhancement|
|Created|2009-09-23 17:33:56Z|
|Branch|https://github.com/twisted/twisted/tree/streaming-zipfile-4035|

As I can see the goal of zipstream is to provide support for an incremental approach to unzipping files. 

But standard zipfile has this support since python 2.6. ZipFile?.open() method returns file-like object which supports 'read' method with buffer size in parameters.

Implementation in zipfile is better than in zipstream. At least it works a lot faster because it don't have 'self.buffer = self.buffer[n:]' code which kills performance on big (even tens of Kb's) buffer sizes like zipstream version does.

I think is better to deprecate zipstream and copy implementation from new zipfile. Or make zipstream a backported zipfile wrapper.

Attachments:

* [4035.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4f2/4f2824d06f8bad7ea2d88cb6a1f3215f847560aa/1af27719699173642d6716125738f31b4a09b3c0.patch) (1589 bytes) - added by aruse on 2011-03-14 17:29:04Z - Modified zippath module to use the standard zipfile module if version is >= 2.6.  Older versions will still use zipstream.
* [4035.patch2](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4f2/4f2824d06f8bad7ea2d88cb6a1f3215f847560aa/92ea050ac977c5ab7801f9f8023f02fabc062799.patch2) (1691 bytes) - added by aruse on 2011-03-14 20:03:14Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4035 4035
type__enhancement enhancement
reporter__kmike kmike
priority__normal normal
milestone__ 
branch__branches_streaming_zipfile_4035 branches/streaming-zipfile-4035
branch_author__lvh lvh
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1253727236000000 1253727236000000
changetime__1317836075000000 1317836075000000
version__None None
owner__lvh lvh
cc__kmike cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

See also [#3666](https://github.com/twisted/twisted/issues/3666).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to kmike|
|-|-|

I agree, it would be good to use the improved implementation on python 2.6 (and using a direct backport is better, because then we only have one bit of code to maintain).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Please note that `ZipPath.open` is the main use-case for this code.  Really, the whole module should have been an implementation detail of `ZipPath` in the first place.

**Comment by twisted-trac** [comment]
|<img alt="aruse's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|aruse set owner to aruse<br>aruse set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="aruse's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|aruse removed owner<br>aruse set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra set owner to aruse|
|-|-|

This patch is missing tests and a news file, see TwistedDevelopment

**Comment by twisted-trac** [comment]
|<img alt="aruse's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|aruse removed owner|
|-|-|

I didn't write any new tests because the existing tests already cover the code that was modified.  I've attached a new patch with a news file.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh set owner to @lvh<br>@lvh set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_10"></a>|@lvh commented|
|-|-|

(For other people watching this: please don't name your N-th patch blah.patchN, because trac no longer recognizes your thing as a patch)

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_11"></a>|@lvh set status to `new`|
|-|-|

Looks great, merging (with previous caveat).

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_12"></a>|@lvh commented|
|-|-|

(In [31050]) Branching to 'streaming-zipfile-4035'

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_13"></a>|@lvh commented|
|-|-|

Buildbot builds: [http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/streaming-zipfile-4035]

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_14"></a>|@lvh set status to `closed`|
|-|-|

(In [31053]) Use python 2.6's streaming zipfile support instead of zipstream

Author: aruse
Reviewer: lvh, thijs
Fixes: [#4035](https://github.com/twisted/twisted/issues/4035)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

[#4498](https://github.com/twisted/twisted/issues/4498) was a duplicate of this.
